pkg/dnsman2/controller/source/common/dnsspecinput.go

```diff
@@ -103,6 +94,69 @@ func GetDNSSpecInputForService(log logr.Logger, state state.AnnotationState, gvk
 	})
 }
 
+// GetDNSSpecInputForIngress gets the DNS spec input for an Ingress resource.
+func GetDNSSpecInputForIngress(log logr.Logger, state state.AnnotationState, gvk schema.GroupVersionKind, ingress *networkingv1.Ingress) (*DNSSpecInput, error) {
+	annotations := GetMergedAnnotation(gvk, state, ingress)
+
+	names, err := getDNSNamesForIngress(log, ingress, annotations)
+	if err != nil {
+		return nil, err
+	}
+	if names == nil {
+		return nil, nil
+	}
+
+	var resolveTargetsToAddresses *bool
+	if v := annotations[dns.AnnotatationResolveTargetsToAddresses]; v != "" {
+		resolveTargetsToAddresses = ptr.To(v == "true")
+	}
+
+	return augmentFromCommonAnnotations(annotations, DNSSpecInput{
+		Names:                     names,
+		Targets:                   getTargetsForIngress(ingress),
+		IPStack:                   annotations[dns.AnnotationIPStack],
+		ResolveTargetsToAddresses: resolveTargetsToAddresses,
+	})
+}
+
+func getDNSNamesForIngress(log logr.Logger, ingress *networkingv1.Ingress, annotations map[string]string) (*utils.UniqueStrings, error) {
+	annotatedNames, err := getDNSNamesFromAnnotations(log, annotations)
+	if err != nil {
+		return nil, err
+	}
+	if annotatedNames == nil {
+		return nil, nil
+	}
+
+	all := annotatedNames.Contains("*")
+	dnsNames := utils.NewUniqueStrings()
+	for _, rule := range ingress.Spec.Rules {
+		host := rule.Host
+		if host != "" && (all || annotatedNames.Contains(host)) {
+			dnsNames.Add(host)
+		}
+	}
+
+	annotatedNames.Remove("*")
+	diff := annotatedNames.Difference(dnsNames)
+	if len(diff) > 0 {
+		return nil, fmt.Errorf("annotated dns names %v not declared by ingress", strings.Join(diff, ", "))
```

Comment:
nit: `%s` is sufficient for string
```suggestion
		return nil, fmt.Errorf("annotated dns names %s not declared by ingress", strings.Join(diff, ", "))
```

Reply:
Done via: [`00b3869` (#739)](https://github.com/gardener/external-dns-management/pull/739/commits/00b3869aa2f515db012add878ee2984c1f2c7fc1)